src/core/socialprovider.js

```diff
@@ -179,19 +179,23 @@ XMPPSocialProvider.prototype.connect = function(continuation) {
     this.vCardStore.updateProperty(this.id, 'status', 'OFFLINE');
   }.bind(this));
   this.client.addListener('close', function(e) {
-    // This may indicate a broken connection to XMPP.
-    // TODO: handle this.
-    this.logger.error('received unhandled close event', e);
+    this.logger.error('received close event', e);
+    if (this.status === 'ONLINE') {
+      // Check if we are still online, otherwise log out.
+      this.ping_();
+    }
   }.bind(this));
   this.client.addListener('end', function(e) {
-    this.logger.error('received end event, status: ' + this.status, e);
-    if (this.status !== 'ONLINE') {
-      // Reject login promise.
+    if (this.status !== 'ONLINE' && this.client) {
+      // Login is still pending, reject the login promise.
+      this.logger.error('Received end event while logging in');
```

Comment:
Yup - that's why it is being moved to underneath ```if (this.status !== 'ONLINE' && this.client) {``` - we get an 'end' event when the user logs out normally, which should not print an error.